File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
# def dfs(triangle, step, idx):
#     if step == len(triangle):
#         return 0
#     return triangle[step][idx] + max(
#         dfs(triangle, step + 1, idx),
#         dfs(triangle, step + 1, idx + 1)
#     )

# def solution(triangle):
#     return triangle[0][0] + max(dfs(triangle, 1, 0), dfs(triangle, 1, 1))

def solution(triangle):
    n = len(triangle)
    dp = [[0] * n for _ in range(n)]
    for i in range(n-1, -1, -1):
        for j in range(i+1):
            if i == n - 1:
                dp[i][j] = triangle[i][j]
            else:
                dp[i][j] = triangle[i][j] + max(dp[i+1][j], dp[i+1][j+1])
    
    return dp[0][0]

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.Arrays;

class Solution {
    public int solution(int[][] triangle) {
        int size = triangle.length;
        int[][] dp = new int[size][size];
        dp[0][0] = triangle[0][0];
        for (int x = 1; x < size; x++) {
            for (int y = 0; y <= x; y++) {
                if (y == 0) {
                    dp[x][y] = triangle[x][y] + dp[x-1][y];
                } else if (y == x) {
                    dp[x][y] = triangle[x][y] + dp[x-1][y-1];
                } else {
                    dp[x][y] = Math.max(dp[x-1][y], dp[x-1][y-1]) + triangle[x][y];
                }
            }
        }
        int answer = Arrays.stream(dp[size-1]).max().getAsInt();
        return answer;
    }
}